frontend/test/lit/test_peephole_optimizations.py

```diff
@@ -0,0 +1,114 @@
+# Copyright 2024 Xanadu Quantum Technologies Inc.
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""
+This file performs the frontend lit tests that the peephole transformations are correctly applied.
+Each test has two components:
+  1. A qnode with a peephole optimization applied, usually called "f"
+  2. The SAME qnode without a peephole optimization applied, usually called "g"
+
+We need to check that:
+  1. For "f", the peephole transform is correctly applied in mlir
+  2. For "g", the peephole transform is correctly not applied in mlir
+  3. "f" and "g" returns the same results.
+"""
+
+# RUN: %PYTHON %s | FileCheck %s
+
+# pylint: disable=line-too-long
+
+import shutil
+
+import numpy as np
+import pennylane as qml
+
+from catalyst import cancel_inverses, qjit
+
+
+def flush_peephole_opted_mlir_to_iostream(filename):
+    """
+    The QJIT compiler does not offer an interface to access an intermediate mlir in the pipeline.
+    The `QJIT.mlir` is the mlir before any passes are run, i.e. the "0_<qnode_name>.mlir".
+    Since the QUANTUM_COMPILATION_PASS is located in the middle of the pipeline, we need
+    to retrieve it with keep_intermediate=True and manually access the "2_QuantumCompilationPass.mlir".
```

Comment:
Is there a better way to do this? i.e. is there a way to access the intermediate stages of mlir in the pipeline?

Reply:
Why do you need the string representation in python if you are doing the pass in MLIR?

Reply:
I think what Erick is saying is (which I would also like to know): Why aren't these tests just in the MLIR test suite, why are they in the frontend?

To answer your question though, the way you're doing it is the mechanism we have for it right now. It's primarily intended for debugging, so the reason there isn't a better way to this for testing is related to the MLIR vs Python question.

Reply:
@dime10 this comment is stale. It was meant more for reading (I think) from the string representation in python, before the transform dialect was used.

Reply:
But we do have an interface to print specific passes. `git grep print_compilation_stage`.

Reply:
Got it, thanks! Yeah `print_compilation_stage` is what's being used here. I still wonder why we are expressing these tests in the Python lit suite rather than in the MLIR suite.

Reply:
@dime10 The tests are here as python lit test because, the entire point of this PR is to give the frontend a way to invoke the cancel inverse pass. The mlir test for `-remove-chained-self-inverse` already exists https://github.com/PennyLaneAI/catalyst/blob/main/mlir/test/Quantum/ChainedSelfInverseTest.mlir (which could use some work imo but that's a seperate issue). 

The tests associated to this PR is really "is the frontend `@catalyst.cancel_inverses` decorator correctly calling the `-remove-chained-self-inverse` pass". Upon second thought, the specific action (aka cancel the Hadamards) of the pass indeed doesn't fall into the jurisdiction of this PR, and is tested already by the mlir tests.

Then, the test needed by this PR is really whether the decorator is correctly lowering to our new jax primitives, and whether the new jax primitives correctly lower to the transform dialect. I do think these two tests are down below in the same file. This first test is more end-to-end, but if we follow Lee's guide and focus on unit tests then I agree the end-to-end test does not need to be there.

Reply:
Fair enough! I think the MLIR tests could be updated to test applying the passes via the transform dialect, and we could have some units here that check the transform instructions are properly generated. 

Having an integration test on top of that is acceptable I suppose :)